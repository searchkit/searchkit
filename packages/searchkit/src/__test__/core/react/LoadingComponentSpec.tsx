import * as React from "react";

import {
  SearchkitManager,
  LoadingComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    this.wrapper = mount(
      <LoadingComponent searchkit={this.searchkit}>
        <p>loading...</p>
      </LoadingComponent>
    )
  })

  it("display nothing when not loading", ()=> {
    expect(this.wrapper).toMatchSnapshot()
  })

  it("display children when loading", ()=> {
    this.searchkit.loading = true
    this.wrapper.instance().forceUpdate()
    expect(this.wrapper).toMatchSnapshot()
  })


})
