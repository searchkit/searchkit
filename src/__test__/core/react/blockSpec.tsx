import {
  block
} from "../../../"

describe("block", ()=> {

  it("block test", ()=> {
    let b = block("foo")
    // console.log(typeof b)
    // console.log(b.toString())
    // console.log(b.mix("bar").state({hover:true}).toString())
    let bemBlocks = {
      item:block("item"),
      container:block("container")
    }

    expect(bemBlocks.container.el("panel").mix("red").state({active:true}).toString())
      .toEqual("container__panel red is-active")

    expect(bemBlocks.container.el("panel").mix(bemBlocks.item.el().mod({mod:"yes"})).state({active:true}).toString())
      .toEqual("container__panel item item_mod_yes is-active")

    expect(bemBlocks.container.mod({type:"text"}).mix("red").state({active:true}).toString())
      .toEqual("container container_type_text red is-active")

    expect(block("button").el("icon").mod({name:"check"}).toString())
      .toEqual("button__icon button__icon_name_check")

    expect(block("button").el().mod({name:"check"}).toString())
      .toEqual("button button_name_check")


    expect(block("button").el().mix(["foo", "bar"]).toString())
      .toEqual("button foo bar")

  })

})
