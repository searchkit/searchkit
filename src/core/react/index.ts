export * from "./SearchkitComponent"
export * from "./SearchkitProvider"
export * from "./LoadingComponent"
export * from "./ReactComponentType"
export * from "./FastClick"
export * from "./pure-render/index"
