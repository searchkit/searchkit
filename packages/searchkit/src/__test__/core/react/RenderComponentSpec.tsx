import * as React from "react";
import * as PropTypes from "prop-types";
import {mount} from "enzyme";
import {
  fastClick
} from "../../../components/__test__/TestHelpers"

import {Panel} from "../../../components"

import {
  renderComponent,
  RenderComponentType,
  RenderComponentPropType
} from "../../../"


describe("RenderComponent", ()=> {

  beforeEach(()=> {
    this.SubPanel = class SubPanel extends Panel {
    }
    this.SubPanel.defaultProps.title = "SubPanel"
    this.SubPanelElement = <Panel title="PanelElement"/>

    this.PanelReactClass = class PanelReactClass extends React.Component {
      static contextTypes = {
        color:PropTypes.string
      }
      render(){
        return (<Panel title={"PanelReactClass " + this.context.color} {...this.props}>
          {this.props.children}
        </Panel>)
      }
    }



    this.PanelFunction = (props, context)=> {
      return (
        <Panel title={"PanelFunction " + context.color}>
          {props.children}
        </Panel>
      )
    }
    this.PanelFunction.contextTypes = {
      color:PropTypes.string
    }

    class Provider extends React.Component<any, any>{
      static childContextTypes = {
        color: PropTypes.string
      }
      getChildContext(){
        return {color:"purple"}
      }
      render(){
        return this.props.children
      }
    }

    this.mount = (component, props={})=> {
      this.wrapper = mount(
        <Provider>
          {renderComponent(
            component, props,
            <p>content..</p>
          )}
        </Provider>
      )
    }

  })

  it("React.Component class", ()=> {
    this.mount(this.SubPanel)
    expect(this.wrapper).toMatchSnapshot()

  })

  it("React Element", ()=> {
    this.mount(this.SubPanelElement)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("React Class", ()=> {
    this.mount(this.PanelReactClass)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Render function", ()=> {
    this.mount(this.PanelFunction)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Invalid component", ()=> {
    spyOn(console, "warn")
    try{
      this.mount(10)
    } catch (e){

    }
    expect(console.warn).toHaveBeenCalledWith(
      "Invalid component", 10
    )

  })

})
