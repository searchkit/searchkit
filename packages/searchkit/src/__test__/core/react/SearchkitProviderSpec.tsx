import * as React from "react";

import {
  SearchkitProvider,
  SearchkitManager,
  SearchkitComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    class SomeComponent extends SearchkitComponent<any, any> {

      render(){
        return <h1>Hello</h1>
      }
    }
    this.wrapper = mount(
      <SearchkitProvider searchkit={this.searchkit}>
        <SomeComponent/>
      </SearchkitProvider>
    )
  })

  it("searchkit provider should work correctly", ()=> {
    expect(this.wrapper.html()).toBe("<h1>Hello</h1>")
    expect(this.wrapper.node.props.searchkit)
      .toBe(this.searchkit)
  })

  it("should call setupListeners()", ()=> {
    spyOn(this.searchkit, "setupListeners")
    expect(this.searchkit.setupListeners).not.toHaveBeenCalled()
    this.wrapper.node.componentWillMount()
    expect(this.searchkit.setupListeners).toHaveBeenCalled()
    this.searchkit.guidGenerator.counter = 10
    this.searchkit.unlistenHistory = jasmine.createSpy("unlisten")
    this.wrapper.node.componentWillUnmount()
    expect(this.searchkit.unlistenHistory).toHaveBeenCalled()
    expect(this.searchkit.guidGenerator.counter).toEqual(0)

  })


})
