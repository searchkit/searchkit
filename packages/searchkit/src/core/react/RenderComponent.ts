import * as React from 'react'
import * as PropTypes from "prop-types"

const omitBy = require("lodash/omitBy")
const isUndefined = require("lodash/isUndefined")

export type RenderFunction = (props?:any, children?:any) => Element
export type Element = React.ReactElement<any>
export type RenderComponentType<P> = React.ComponentClass<P> | React.ClassicComponentClass<P> | Element | RenderFunction | any

export const RenderComponentPropType = PropTypes.oneOfType([
  function(props:any, propName: string) {
     if(isUndefined(props[propName]) || (props[propName]["prototype"] instanceof React.Component)) {
       return null
     }
  },
  PropTypes.element,
  PropTypes.func,
])

export function renderComponent(component:RenderComponentType<any>, props={}, children=null): React.ReactElement<any> {
  let isReactComponent = (
    component["prototype"] instanceof React.Component ||
    (component["prototype"] && component["prototype"].isReactComponent) ||
    typeof component === 'function'
  )
  if (isReactComponent){
    return React.createElement(
      component as React.ComponentClass<any>,
      props, children
    )
  } else if (React.isValidElement(component)){
    return React.cloneElement(
      component as Element,
      omitBy(props, isUndefined), children
    );
  }
  console.warn("Invalid component", component)
  return null
}
