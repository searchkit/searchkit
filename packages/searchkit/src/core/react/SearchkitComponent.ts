import * as React from "react"
import * as PropTypes from "prop-types"
import {SearchkitManager} from "../SearchkitManager"
import {ImmutableQuery} from "../query"
import {Accessor} from "../accessors/Accessor"
import {Utils} from "../support"
const mapValues = require("lodash/mapValues")
import {block} from "./block"

export interface SearchkitComponentProps {
  mod?:string
  className?:string
  translations?:Object
  searchkit?:SearchkitManager
  key?:string
}

export class SearchkitComponent<P extends SearchkitComponentProps,S> extends React.Component<P,S> {
  _searchkit:SearchkitManager
  accessor:Accessor
  stateListenerUnsubscribe:Function
  translations:Object = {}
  unmounted = false

  static contextTypes: React.ValidationMap<any> = {
		searchkit:PropTypes.instanceOf(SearchkitManager)
	}

  static translationsPropType = (translations)=> {
    return PropTypes.shape(mapValues(translations, ()=> PropTypes.string))
  }

  static propTypes:any = {
    mod :PropTypes.string,
    className :PropTypes.string,
    translations: PropTypes.objectOf(
      PropTypes.string),
    searchkit:PropTypes.instanceOf(SearchkitManager)
  }

  constructor(props?){
    super(props)
    this.translate = this.translate.bind(this)
  }

  defineBEMBlocks() {
    return null;
  }

  defineAccessor():Accessor{
    return null
  }

  translate(key, interpolations?){
    let translation = (
      (this.searchkit.translate(key)) ||
      (this.props.translations && this.props.translations[key]) ||
      this.translations[key] || key)
    return Utils.translate(translation, interpolations)
  }

  get bemBlocks(): any {
    return mapValues(this.defineBEMBlocks(), (cssClass) => {
      return block(cssClass).el
    })
  }

  get searchkit(): SearchkitManager {
    return this._searchkit || (this._searchkit = this._getSearchkit());
  }

  _getSearchkit(): SearchkitManager {
    return this.props.searchkit || this.context["searchkit"];
  }

  set searchkit(value: SearchkitManager) {
    this._searchkit = value;
  }

  componentDidMount() {
    this._initAccessor();
    if (this.searchkit) {
      this.stateListenerUnsubscribe = this.searchkit.emitter.addListener(() => {
        if (!this.unmounted) {
          this.forceUpdate();
        }
      });
    }
    this.forceUpdate();
  }
  
  _initAccessor() {
    if (this.searchkit && !this.accessor) {
      this.accessor = this.defineAccessor();
      if (this.accessor) {
        this.accessor = this.searchkit.addAccessor(this.accessor);
        return true;
      }
    }
    if (!this.searchkit) {
      console.warn(
        "No searchkit found in props or context for " + this.constructor["name"]
      );
    }
    return false;
  }

  componentWillUnmount(){
    if(this.stateListenerUnsubscribe){
		  this.stateListenerUnsubscribe()
    }
    if(this.searchkit && this.accessor){
      this.searchkit.removeAccessor(this.accessor)
    }
    this.unmounted = true
	}

  getResults(){
    return this.searchkit.results
  }

  getHits(){
    return this.searchkit.getHits()
  }

  getHitsCount(){
    return this.searchkit.getHitsCount()
  }

  hasHits(){
    return this.searchkit.hasHits()
  }

  hasHitsChanged(){
    return this.searchkit.hasHitsChanged()
  }

  getQuery():ImmutableQuery {
    return this.searchkit.query
  }

  isInitialLoading(){
    return this.searchkit.initialLoading
  }

  isLoading(){
    return this.searchkit.loading
  }

  getError(){
    return this.searchkit.error
  }


}
