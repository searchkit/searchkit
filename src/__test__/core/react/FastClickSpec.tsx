import * as React from "react";

import {
  SearchkitManager,
  FastClick
} from "../../../"

import {mount} from "enzyme";

describe("FastClick", ()=> {

  beforeEach(()=> {
    this.handler = jasmine.createSpy("fastclick handler")
    this.wrapper = mount(
      <FastClick handler={this.handler}>
        <button>click me</button>
      </FastClick>
    )
  })

  it("should render children", ()=> {
    expect(this.wrapper.html())
      .toEqual("<button>click me</button>")
  })

  it("test mousedown", ()=> {
    this.wrapper.simulate("mouseDown", {button:1})
    expect(this.handler).not.toHaveBeenCalled()

    this.wrapper.simulate("mouseDown", {button:0})
    expect(this.handler).toHaveBeenCalled()
  })

  describe("Touch events", ()=> {

    beforeEach(()=> {
      this.fastClick = this.wrapper.node
      this.simulateTouch = (event, x, y)=> {
        this.wrapper.simulate(event, {
          changedTouches:[ {pageX:x, pageY:y} ]
        })
      }

      this.simulateTouch("touchStart", 10, 20)
      expect(this.fastClick.supportsTouch).toBe(true)
      expect(this.fastClick.startPoint).toEqual({
        x:10, y:20
      })

    })

    it("test touch above threshold", ()=> {
      this.simulateTouch("touchEnd", 30, 40)
      expect(this.handler).not.toHaveBeenCalled()
      expect(this.fastClick.startPoint).toBe(undefined)

      //ignore mousedowns if supports touch
      this.wrapper.simulate("mouseDown", {button:0})
      expect(this.handler).not.toHaveBeenCalled()
    })


    it("test touch within threshold", ()=> {
      this.simulateTouch("touchEnd", 29, 39)
      expect(this.handler).toHaveBeenCalled()
      expect(this.fastClick.startPoint).toBe(undefined)
    })

    it("ignore multiple touch", ()=> {
      this.wrapper.simulate("touchStart", {
        changedTouches:[
          {pageX:1,pageY:2},
          {pageX:10,pageY:22}
        ]
      })
      expect(this.fastClick.startPoint).toBe(null)
    })

    it("prevents default click behaviour", ()=> {
      let event = {
        preventDefault:jasmine.createSpy("preventDefault")
      }
      this.wrapper.simulate("click", event)
      expect(event.preventDefault).toHaveBeenCalled()
    })

  })


})
