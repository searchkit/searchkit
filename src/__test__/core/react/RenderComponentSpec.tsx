import * as React from "react";
import {mount} from "enzyme";
import {
  fastClick, hasClass, jsxToHTML, printPrettyHtml
} from "../../../components/__test__/TestHelpers"

import {Panel} from "../../../components"

import {
  renderComponent,
  RenderComponentType,
  RenderComponentPropType
} from "../../../"


describe("RenderComponent", ()=> {

  beforeEach(()=> {
    this.SubPanel = class SubPanel extends Panel {
    }
    this.SubPanel.defaultProps.title = "SubPanel"
    this.SubPanelElement = <Panel title="PanelElement"/>
    this.PanelReactClass = React.createClass({
      contextTypes:{
        color:React.PropTypes.string
      },
      render(){
        return (<Panel title={"PanelReactClass " + this.context.color} {...this.props}>
          {this.props.children}
        </Panel>)
      }
    })

    this.PanelFunction = (props, context)=> {
      return (
        <Panel title={"PanelFunction " + context.color}>
          {props.children}
        </Panel>
      )
    }
    this.PanelFunction.contextTypes = {
      color:React.PropTypes.string
    }

    class Provider extends React.Component<any, any>{
      static childContextTypes = {
        color: React.PropTypes.string
      }
      getChildContext(){
        return {color:"purple"}
      }
      render(){
        return this.props.children
      }
    }

    this.mount = (component, props={})=> {
      this.wrapper = mount(
        <Provider>
          {renderComponent(
            component, props,
            <p>content..</p>
          )}
        </Provider>
      )
    }

  })

  it("React.Component class", ()=> {
    this.mount(this.SubPanel)
    expect(this.wrapper.html()).toEqual(jsxToHTML(
      <div className="sk-panel">
        <div className="sk-panel__header">SubPanel</div>
        <div className="sk-panel__content">
          <p>content..</p>
        </div>
      </div>
    ))
  })

  it("React Element", ()=> {
    this.mount(this.SubPanelElement)
    expect(this.wrapper.html()).toEqual(jsxToHTML(
      <div className="sk-panel">
        <div className="sk-panel__header">PanelElement</div>
        <div className="sk-panel__content">
          <p>content..</p>
        </div>
      </div>
    ))
  })

  it("React Class", ()=> {
    this.mount(this.PanelReactClass)
    expect(this.wrapper.html()).toEqual(jsxToHTML(
      <div className="sk-panel">
        <div className="sk-panel__header">PanelReactClass purple</div>
        <div className="sk-panel__content">
          <p>content..</p>
        </div>
      </div>
    ))
  })

  it("Render function", ()=> {
    this.mount(this.PanelFunction)
    expect(this.wrapper.html()).toEqual(jsxToHTML(
      <div className="sk-panel">
        <div className="sk-panel__header">PanelFunction purple</div>
        <div className="sk-panel__content">
          <p>content..</p>
        </div>
      </div>
    ))
  })

  it("Invalid component", ()=> {
    spyOn(console, "warn")
    try{
      this.mount(10)
      // printPrettyHtml(this.wrapper.html())
    } catch (e){

    }
    expect(console.warn).toHaveBeenCalledWith(
      "Invalid component", 10
    )

  })

})
