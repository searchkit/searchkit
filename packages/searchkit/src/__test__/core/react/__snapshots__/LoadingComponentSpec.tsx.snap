// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchkitProvider display children when loading 1`] = `<p>loading...</p>`;

exports[`SearchkitProvider display nothing when not loading 1`] = `<div></div>`;
